### PR TITLE
doc: clarify that LOCK() internally checks whether the mutex is held

### PR DESCRIPTION
**Pull request by vasild**
Constructs like

```cpp
AssertLockNotHeld(m);
LOCK(m);
```

are equivalent to (almost, modulo some logging differences, see below)

```cpp
LOCK(m);
```

for non-recursive mutexes, so it is ok to omit `AssertLockNotHeld()` in such cases. Requests to do the former keep coming during review process. `developer-notes.md` explicitly states "Combine annotations in function declarations with run-time asserts in function definitions", but that seems to be too strong or unclear. `LOCK()` is also a run-time assert in this case.

Also remove `LocksHeld()` from the public interface in `sync.h` since it is only used in `sync.cpp`.

### CODE REVIEW

**Comment by ajtowns** [comment]
These aren't precisely the same (maybe they should be?). With `AssertLockNotHeld` you just get a report on that lock, on stderr:

```
Assertion failed: lock flock held in net.cpp:1398; locks held:
'flock' in net.cpp:1404 (in thread 'dnsseed')
```

with a bare `LOCK()` you get a less useful report on stderr (it only shows the location of the second lock attempt, not where it's being held):

```
Assertion failed: detected double lock for 'flock' in net.cpp:1399 (in thread 'dnsseed'), details in debug log.
```

but get more information in debug.log:

```
2023-02-17T11:47:18Z DOUBLE LOCK DETECTED
2023-02-17T11:47:18Z Lock order:
2023-02-17T11:47:18Z  (*) 'flock' in net.cpp:1404 (in thread 'dnsseed')
2023-02-17T11:47:18Z  (*) 'flock' in net.cpp:1399 (in thread 'dnsseed')
```

**Comment by vasild** [comment]
Do you think that this is a blocker for this PR?

**Comment by ajtowns** [comment]
> Do you think that this is a blocker for this PR?

I was just intending to note the difference not express an opinion (which I don't really have at this point). Would be nicer if the approach this PR recommends was clearly better in every way, though.

File: doc/developer-notes.md [hunk 1]
```diff
@@ -941,7 +941,9 @@ Threads and synchronization
     internal to a class (private or protected) rather than public.
 
   - Combine annotations in function declarations with run-time asserts in
-    function definitions:
+    function definitions (`AssertLockNotHeld()` can be omitted if `LOCK()` is
+    called unconditionally after it because `LOCK()` does the same check as
```
Comment (hebasto) [inline doc/developer-notes.md:945]: > ... can be omitted if \<condition\>

It creates additional burden for reviewing if \<condition\> will change in the future, doesn't it?

**Review by hebasto — COMMENTED** [review]
Agree with https://github.com/bitcoin/bitcoin/pull/27116#issuecomment-1434539481.

File: doc/developer-notes.md (hunk 1 above)
Comment (ajtowns) [reply doc/developer-notes.md:945]: If you're moving locks around (`<condition>` changes), then you should be reviewing whether `AssertLockNotHeld` is still appropriate either way?

Comment (vasild) [reply doc/developer-notes.md:945]: Yes, if `LOCK()` is moved, then it would have to be reviewed either way.

Comment (Ely1969) [reply doc/developer-notes.md:945]: Hoe to merge

**Comment by vasild** [comment]
1. Should I change `LOCK()` to print everything to `stderr`, in addition to the log? That is, print this to `stderr` as well:

```
2023-02-17T11:47:18Z DOUBLE LOCK DETECTED
2023-02-17T11:47:18Z Lock order:
2023-02-17T11:47:18Z  (*) 'flock' in net.cpp:1404 (in thread 'dnsseed')
2023-02-17T11:47:18Z  (*) 'flock' in net.cpp:1399 (in thread 'dnsseed')
```

2. Or do people prefer to write both `AssertLockNotHeld()` and `LOCK()`, like this:

```cpp
void f()
{
    AssertLockNotHeld(mutex);
    LOCK(mutex);
    ...
```

If "yes", then I will close this PR.

3. Or call `AssertLockNotHeld(mutex)` from inside `LOCK()` if the mutex is not recursive?

**Comment by ajtowns** [comment]
I lean lightly towards (1), but prefer "(0) keep this PR as is" over (2) or (3).

**Review by hebasto — APPROVED** [review]
ACK 91d08889218e06631f43a3dab0bae576aa46e43c, I have reviewed the code and it looks OK.

> I lean lightly towards (1), but prefer "(0) keep this PR as is" over (2) or (3).

My vote is for (0) or (1).

**Comment by achow101** [comment]
ACK 91d08889218e06631f43a3dab0bae576aa46e43c

I think it's fine to have people have to look at the debug.log for additional details. Running with `-daemon` results in stderr being lost anyways.

**Pull request merged**